flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -507,12 +509,18 @@ public void suspendJob(final Throwable cause) {
 	 * @return Acknowledge the task execution state update
 	 */
 	@RpcMethod
-	public boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
+	public Acknowledge updateTaskExecutionState(final TaskExecutionState taskExecutionState) throws ExecutionGraphException {
 		if (taskExecutionState == null) {
-			return false;
+			throw new NullPointerException("TaskExecutionState must not be null.");
+		}
+
+		if (executionGraph.updateState(taskExecutionState)) {
+			return Acknowledge.get();
 		} else {
-			return executionGraph.updateState(taskExecutionState);
+			throw new ExecutionGraphException("The execution attempt " +
```

Comment:
Execution attempt not exists is one of the failure causes, maybe we can move the throw behavior into the function call updateState, and let the function decide which exception should be thrown